internal/pkg/gateway/api.go

```diff
@@ -161,16 +163,14 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
 		var errDetails []proto.Message
 		for firstResponse == nil && firstEndorser != nil {
 			firstResponse, err = firstEndorser.client.ProcessProposal(ctx, signedProposal)
-			if err != nil {
-				logger.Debugw("Endorse call to endorser failed", "channel", request.ChannelId, "txid", request.TransactionId, "endorserAddress", firstEndorser.endpointConfig.address, "endorserMspid", firstEndorser.endpointConfig.mspid, "error", err)
-				errDetails = append(errDetails, errorDetail(firstEndorser.endpointConfig, err))
-				gs.registry.removeEndorser(firstEndorser)
-				firstEndorser = plan.retry(firstEndorser)
-			} else if firstResponse.Response.Status < 200 || firstResponse.Response.Status >= 400 {
-				logger.Debugw("Endorse call to endorser returned failure", "channel", request.ChannelId, "txid", request.TransactionId, "endorserAddress", firstEndorser.endpointConfig.address, "endorserMspid", firstEndorser.endpointConfig.mspid, "status", firstResponse.Response.Status, "message", firstResponse.Response.Message)
-				err := fmt.Errorf("error %d, %s", firstResponse.Response.Status, firstResponse.Response.Message)
-				endpointErr := errorDetail(firstEndorser.endpointConfig, err)
-				errDetails = append(errDetails, endpointErr)
+			success, message, _, remove := gs.responseStatus(firstResponse, err)
+
+			if !success {
+				logger.Debugw("Endorse call to endorser failed", "channel", request.ChannelId, "txid", request.TransactionId, "endorserAddress", firstEndorser.endpointConfig.address, "endorserMspid", firstEndorser.endpointConfig.mspid, "error", message)
+				errDetails = append(errDetails, errorDetail(firstEndorser.endpointConfig, message))
+				if remove {
+					gs.registry.removeEndorser(firstEndorser)
+				}
 				firstEndorser = plan.retry(firstEndorser)
```

Comment:
What is the meaning of this? Are we retrying on the first endorser again?

Reply:
We're retrying to get the first endorsement (the one that generates the interest) from another peer in the same org

Reply:
Ok, it is not obvious from reading the code, maybe change retry() function name to something like retryOnDifferentPeerInOrg().